KAFKA-16373: KIP-1028: Adding 3.7.0 docker official images static assets (#1)

KrishVora01 opened the pull request:
This PR aims to add the static Dockerfile and scripts for AK 3.7.0 version. As mentioned in [KIP-1028's](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1028%3A+Docker+Official+Image+for+Apache+Kafka#KIP1028:DockerOfficialImageforApacheKafka-ProposedWorkflowforReleaseProcess) discussion thread [here](https://lists.apache.org/thread/pfk2hqssbkt5ffblnh0gs7fw4w8ldcz6), since this KIP was voted, this PR aims to start the release of the kafka:3.7.0 Docker Official image. This will also help us validate the process and allow us to address any changes suggested by Dockerhub before the 3.8.0 release.

The static Dockerfile and scripts have been generated via the github actions workflows and scripts added as part of [PR-16027](https://github.com/apache/kafka/pull/16027). The reports of build and testing the 3.7.0 Docker official image are below.
[report_jvm.html.zip](https://github.com/user-attachments/files/15613529/report_jvm.html.zip)
[scan_report_jvm.txt.zip](https://github.com/user-attachments/files/15613530/scan_report_jvm.txt.zip)

**Post this PR:**
1. Run the [docker/generate_kafka_pr_template.py](https://github.com/apache/kafka/blob/trunk/docker/generate_kafka_pr_template.py) script, and raise the corresponding PR in [Docker Official Images repo](https://github.com/docker-library/official-images/tree/master/library).

**Committer Checklist (excluded from commit message)**
- [ ]  Verify design and implementation
- [ ]  Verify test coverage and CI build status

KrishVora01 committed:
Adding 3.7.0 docker official images static assets (#1)

loicgreffier committed:
KAFKA-16448: Add ProcessingExceptionHandler interface and implementations (#16187)

This PR is part of KAFKA-16448 which aims to bring a ProcessingExceptionHandler to Kafka Streams in order to deal with exceptions that occur during processing.

This PR brings ProcessingExceptionHandler interface and default implementations.

Co-authored-by: Dabz <d.gasparina@gmail.com>
Co-authored-by: sebastienviale <sebastien.viale@michelin.com>

Reviewer: Bruno Cadonna <cadonna@apache.org>

sjhajharia committed:
MINOR: Code cleanup in metadata module (#16065)


Reviewers: Mickael Maison <mickael.maison@gmail.com>

caysony1 committed:
MINOR: Added test for ClusterConfig#displayTags (#16110)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

kamalcph committed:
MINOR: Cleanup the storage module unit tests (#16202)

- Use SystemTime instead of MockTime when time is not mocked
- Use static assertions to reduce the line length
- Fold the lines if it exceeds the limit
- rename tp0 to tpId0 when it refers to TopicIdPartition

Reviewers: Kuan-Po (Cooper) Tseng <brandboat@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

muralibasani committed:
KAFKA-16884 Refactor RemoteLogManagerConfig with AbstractConfig (#16199)

Reviewers: Greg Harris <gharris1727@gmail.com>, Kamal Chandraprakash <kamal.chandraprakash@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

mimaison committed:
MINOR: Various cleanups in clients (#16193)


Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

dajac committed:
KAFKA-14701; Move `PartitionAssignor` to new `group-coordinator-api` module (#16198)

This patch moves the `PartitionAssignor` interface and all the related classes to a newly created `group-coordinator/api` module, following the pattern used by the storage and tools modules.

Reviewers: Ritika Reddy <rreddy@confluent.io>, Jeff Kim <jeff.kim@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

cadonna committed:
KAFKA-16903: Consider produce error of different task (#16222)

A task does not know anything about a produce error thrown
by a different task. That might lead to a InvalidTxnStateException
when a task attempts to do a transactional operation on a producer
that failed due to a different task.

This commit stores the produce exception in the streams producer
on completion of a send instead of the record collector since the
record collector is on task level whereas the stream producer
is on stream thread level. Since all tasks use the same streams
producer the error should be correctly propagated across tasks
of the same stream thread.

For EOS alpha, this commit does not change anything because
each task uses its own producer. The send error is still
on task level but so is also the transaction.

Reviewers: Matthias J. Sax <matthias@confluent.io>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 19) Update to new AssignmentConfigs (#16219)

This PR updates all of the streams task assignment code to use the new AssignmentConfigs public class.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

apoorvmittal10 committed:
KAFKA-16905: Fix blocking DescribeCluster call in AdminClient DescribeTopics (#16217)

Reviewers: Andrew Schofield <aschofield@confluent.io>, Lianet Magrans <lianetmr@gmail.com>, David Arthur <mumrah@gmail.com>

johnnychhsu committed:
KAFKA-16223 Migrate KafkaConfigBackingStoreTest to Mockito (#16126)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

brenden20 committed:
KAFKA-16912 Migrate ConsumerNetworkThreadTest.testPollResultTimer() to NetworkClientDelegateTest (#16234)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

kirktrue committed:
KAFKA-16200: Enforce that RequestManager implementations respect user-provided timeout (#16031)

Improve consistency and correctness for user-provided timeouts at the Consumer network request layer, per the Java client Consumer timeouts design (https://cwiki.apache.org/confluence/display/KAFKA/Java+client+Consumer+timeouts). While the changes introduced in KAFKA-15974 enforce timeouts at the Consumer's event layer, this change enforces timeouts at the network request layer.

The changes mostly fit into the following areas:

1. Create shared code and idioms so timeout handling logic is consistent across current and future RequestManager implementations
2. Use deadlineMs instead of expirationMs, expirationTimeoutMs, retryExpirationTimeMs, timeoutMs, etc.
3. Update "preemptive pruning" to remove expired requests that have had at least one attempt

Reviewers: Lianet Magrans <lianetmr@gmail.com>, Bruno Cadonna <cadonna@apache.org>

soarez committed:
KAFKA-16606 Gate JBOD configuration on 3.7-IV2 (#15834)

Support for multiple log directories in KRaft exists from
MetataVersion 3.7-IV2.

When migrating a ZK broker to KRaft, we already check that
the IBP is high enough before allowing the broker to startup.

With KIP-584 and KIP-778, Brokers in KRaft mode do not require
the IBP configuration - the configuration is deprecated.
In KRaft mode inter.broker.protocol.version defaults to
MetadataVersion.MINIMUM_KRAFT_VERSION (IBP_3_0_IV1).

Instead KRaft brokers discover the MetadataVersion by reading
the "metadata.version" FeatureLevelRecord from the cluster metadata.

This change adds a new configuration validation step upon discovering
the "metadata.version" from the cluster metadata.

Reviewers: Mickael Maison <mickael.maison@gmail.com>

ijuma committed:
MINOR: Remove unnecessary log4j-appender dependency and tweak explicit log4j dependency (#12148)


Reviewers: Mickael Maison <mickael.maison@gmail.com>

Phuc-Hong-Tran committed:
KAFKA-16493: Avoid unneeded subscription regex check if metadata version unchanged (#15869)

This PR includes changes for AsyncKafkaConsumer to avoid evaluating the subscription regex on every poll if metadata hasn't changed. The metadataVersionSnapshot was introduced to identify whether metadata has changed or not, if yes then the current subscription regex will be evaluated.

This is the same mechanism used by the LegacyConsumer.

Reviewers: Lianet Magrans <lianetmr@gmail.com>, Matthias J. Sax <matthias@confluent.io>

JimGalasyn committed:
KAFKA-16911: Update docs for KIP-862 (#16246)

Reviewers: Matthias J. Sax <matthias@confluent.io>

sebastienviale committed:
MINOR: update all-latency-max, range-latency-avg|max and add prefix-scan documentation (#16182)

Reviewers: Matthias J. Sax <matthias@confluent.io>

JimGalasyn committed:
MINOR: Update docs for for KIP-671 (#16247)

Reviewers: Matthias J. Sax <matthias@confluent.io>

nicolasguyomar committed:
MINOR: Remove AK 1.0.0 reference from NetworkClient.java (#16223)

Reviewers: Justine Olshan <jolshan@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

ocadaruma committed:
KAFKA-16916: Disable blocking test temporarily (#16248)

The test runs forever. We disable the test temporarily to unblock CI

Reviewers: Luke Chen <showuon@gmail.com>

apoorvmittal10 committed:
KAFKA-16916: Fixing error in completing future (#16249)

Fix to complete Future which was stuck due the exception.getCause() throws an error.

The fix in the #16217 unblocked blocking thread but exception in catch block from blocking get call was wrapped in ExecutionException which is not the case when moved to async workflow hence getCause is not needed.

Reviewers: Luke Chen <showuon@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

soarez committed:
MINOR: Fix broken ReassignPartitionsCommandTest test (#16251)

KAFKA-16606 (#15834) introduced a change that broke
ReassignPartitionsCommandTest.testReassignmentCompletionDuringPartialUpgrade.

The point was to validate that the MetadataVersion supports JBOD
in KRaft when multiple log directories are configured.
We do that by checking the version used in
kafka-features.sh upgrade --metadata, and the version discovered
via a FeatureRecord for metadata.version in the cluster metadata.

There's no point in checking inter.broker.protocol.version in
KafkaConfig, since in KRaft, that configuration is deprecated
and ignored — always assuming the value of MINIMUM_KRAFT_VERSION.

The broken that was broken sets inter.broker.protocol.version in
KRaft mode and configures 3 directories. So alternatively, we
could change the test to not configure this property.
Since the property isn't forbidden in KRaft mode, just ignored,
and operators may forget to remove it, it seems better to remote
the fail condition in KafkaConfig.

Reviewers: Luke Chen <showuon@gmail.com>

soarez committed:
KAFKA-16886: Detect replica demotion in AssignmentsManager (#16232)

JBOD Brokers keep the Controller up to date with replica-to-directory
placement via AssignReplicasToDirsRequest. These requests are queued,
compacted and sent by AssignmentsManager.

The Controller returns the error NOT_LEADER_OR_FOLLOWER when handling
a AssignReplicasToDirsRequest from a broker that is not a replica.

A partition reassignment can take place, removing the Broker
as a replica before the AssignReplicasToDirsRequest successfully
reaches the Controller. AssignmentsManager retries failed
requests, and will continuously try to propagate this assignment,
until the Broker either shuts down, or is added back as a replica.

When encountering a NOT_LEADER_OR_FOLLOWER error, AssignmentsManager
should assume that the broker is no longer a replica, and stop
trying to propagate the directory assignment for that partition.

Reviewers: Luke Chen <showuon@gmail.com>

FrankYang0529 committed:
KAFKA-16878 Remove powermock and easymock from code base (#16236)

Reviewers: TaiJuWu <tjwu1217@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

gongxuanzhang committed:
KAFKA-10787 Apply spotless to transaction-coordinator and server-common (#16172)

Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

gaurav-narula committed:
KAFKA-16920: close BrokerLifecycleManager in tests (#16252)

Tests in BrokerLifecycleManagerTest do not close BrokerLifecycleManager
if an assertion fails.

This change makes BrokerLifecycleManager an instance variable that is
closed in an `@AfterEach` method.

Reviewers: Igor Soarez <i@soarez.me>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 18) Better assignment testing (#16201)

Added more testing for the StickyTaskAssignor, which includes a large-scale test with rack aware enabled.

Also added a no-op change to StreamsAssignmentScaleTest.java to allow for rack aware optimization testing.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

apourchet committed:
KAFKA-15045: (KIP-924 pt. 20) Custom task assignment configuration fix (#16245)

The StreamsConfig class was not parsing the new task assignment
configuration flag properly, which made it impossible to properly
configure a custom task assignor.

This PR fixes this and adds a bit of INFO logging to help users diagnose
assignor misconfiguration issues.

Reviewers: Anna Sophie Blee-Goldman <ableegoldman@apache.org>

chiacyu committed:
KAFKA-16885 Renamed the enableRemoteStorageSystem to isRemoteStorageSystemEnabled (#16256)

Reviewers: Kamal Chandraprakash <kchandraprakash@uber.com>, Chia-Ping Tsai <chia7712@gmail.com>

VedarthConfluent commented on docker/docker_official_images/3.7.0/jvm/Dockerfile:
```diff
@@ -0,0 +1,95 @@
+USER root
+
+# Get kafka from https://archive.apache.org/dist/kafka and pass the url through build arguments
+ENV kafka_url https://downloads.apache.org/kafka/3.7.0/kafka_2.13-3.7.0.tgz
```
This is generated through a script, but let's either update the comment or update the domain used in the scripts as right now it seems inconsistent.

KrishVora01 committed:
modifying Readme for Docker official images (#16226)

This PR aims to modify the README file under /docker, to include the steps to release the Docker Official Images.

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Vedarth Sharma <vesharma@confluent.io>

KrishVora01 committed:
modifying comments in doi dockerfile (#2)

KrishVora01 committed:
Modifying dockerfile comments (#3)

KrishVora01 committed:
[MINOR] KAFKA-16373: KIP-1028: Modifying dockerfile comments (#16261)

This PR aims to add the replace the dockerfile comments under the jvm/dockerfile  Dockerfile, with updated url.
The original comment read
# Get kafka from https://archive.apache.org/dist/kafka and pass the url through build arguments . For DOI dockerfiles, we replace this with
# Get Kafka from https://downloads.apache.org/kafka, url passed as env var, for version {kafka_version} .

Reviewers: Manikumar Reddy <manikumar.reddy@gmail.com>, Vedarth Sharma <vesharma@confluent.io>

moscicky committed:
MINOR: Add fast_commit ext4 config to ops docs (#15496)


Reviewers: Mickael Maison <mickael.maison@gmail.com>
Co-authored-by: Piotr Rzysko <piotr.rzysko@allegro.com>

tinaselenge committed:
KAFKA-15718: Refactor UncleanLeaderElectionTest to enable KRaft later (#16157)

Refactor UncleanLeaderElectionTest to allow to enable KRaft later

Reviewers: Luke Chen <showuon@gmail.com>

loicgreffier committed:
KAFKA-16448: Add ProcessingExceptionHandler in Streams configuration (#16188)

This PR is part of KAFKA-16448 which aims to bring a ProcessingExceptionHandler to Kafka Streams in order to deal with exceptions that occur during processing.

This PR brings ProcessingExceptionHandler in Streams configuration.

Co-authored-by: Dabz <d.gasparina@gmail.com>
Co-authored-by: sebastienviale <sebastien.viale@michelin.com>

Reviewer: Bruno Cadonna <cadonna@apache.org>

riedelmax committed:
KAFKA-14509; [4/4] Handle includeAuthorizedOperations in ConsumerGroupDescribe API (#16158)

This patch implements the handling of `includeAuthorizedOperations` flag in the ConsumerGroupDescribe API.

Reviewers: David Jacot <djacot@confluent.io>

mimaison committed:
MINOR: Improve documentation of default.replication.factor config (#16243)


Reviewers: Chia-Ping Tsai <chia7712@gmail.com>

dajac committed:
MINOR: Rename uniform assignor's internal builders (#16233)

This patch renames the uniform assignor's builders to match the `SubscriptionType` which is used to determine which one is called. It removes the abstract class `AbstractUniformAssignmentBuilder` which is not necessary anymore. It also applies minor refactoring.

Reviewers: Ritika Reddy <rreddy@confluent.io>, Chia-Ping Tsai <chia7712@gmail.com>

KrishVora01 committed:
Merge remote-tracking branch 'origin/trunk' into trunk

KrishVora01 committed:
adding 3.7.0
